FprimeZephyrReference/Components/Authenticate/Authenticate.cpp

```diff
@@ -0,0 +1,532 @@
+    }
+
+    if (status != PSA_SUCCESS) {
+        U32 statusU32 = static_cast<U32>(status);
+        this->log_WARNING_HI_CryptoComputationError(statusU32);
+        return Fw::Buffer();
+    }
+
+    // Compute outer hash: H(k XOR opad || innerHash)
+    std::vector<U8> outerKey(blockSize);
+    for (size_t i = 0; i < blockSize; i++) {
+        outerKey[i] = preparedKey[i] ^ opad;
+    }
+
+    psa_hash_operation_t outerHash = PSA_HASH_OPERATION_INIT;
+    status = psa_hash_setup(&outerHash, PSA_ALG_SHA_256);
+    if (status == PSA_SUCCESS) {
+        status = psa_hash_update(&outerHash, outerKey.data(), blockSize);
+    }
+    if (status == PSA_SUCCESS) {
+        status = psa_hash_update(&outerHash, innerHashOutput, innerHashLen);
+    }
+    if (status == PSA_SUCCESS) {
+        status = psa_hash_finish(&outerHash, macOutput, sizeof(macOutput), &macOutputLength);
+    }
+
+    if (status != PSA_SUCCESS) {
+        U32 statusU32 = static_cast<U32>(status);
+        this->log_WARNING_HI_CryptoComputationError(statusU32);
+        return Fw::Buffer();
+    }
+
+    // Allocate a buffer for the HMAC result (we only need 16 bytes per CCSDS spec)
+    // But we computed 32 bytes (full SHA-256), so we'll take the first 16
+    const size_t hmacOutputLength = 16;  // CCSDS 355.0-B-2 specifies 16 bytes
+    U8* hmacData = new U8[hmacOutputLength];
+    if (hmacData == nullptr) {
+        return Fw::Buffer();
+    }
+    std::memcpy(hmacData, macOutput, hmacOutputLength);
+
+    // Create Fw::Buffer with the HMAC data (context 0 for now)
+    return Fw::Buffer(hmacData, static_cast<FwSizeType>(hmacOutputLength), 0);
+}
+
+bool Authenticate::validateSequenceNumber(U32 received, U32 expected) {
+    // validate the sequence number by checking if it is within the window of the expected sequence number
+    Fw::ParamValid valid;
+    U32 window = paramGet_SEQ_NUM_WINDOW(valid);
+    /*
+     * Compute the difference between received and expected sequence numbers using unsigned
+     * 32-bit arithmetic. This handles wraparound correctly due to the well-defined behavior
+     * of unsigned integer overflow in C++. For example, if expected=0xFFFFFFFE and received=1,
+     * then (received - expected) == 3 (modulo 2^32). This is a standard technique for
+     * sequence number window validation (see RFC 1982: Serial Number Arithmetic).
+     */
+    const U32 delta = received - expected;
+    if (delta > window) {
+        this->log_WARNING_HI_SequenceNumberOutOfWindow(received, expected, window);
+        return false;
+    }
+    return true;
+}
+
+bool Authenticate::compareHMAC(const U8* expected, const U8* actual, FwSizeType length) const {
+    if (expected == nullptr || actual == nullptr) {
+        return false;
+    }
+    return std::memcmp(expected, actual, static_cast<size_t>(length)) == 0;
+}
+
+void Authenticate ::dataIn_handler(FwIndexType portNum, Fw::Buffer& data, const ComCfg::FrameContext& context) {
```

Comment:
I would split this into 3 validation functions for readability. Maybe something like 
```suggestion
void Authenticate ::dataIn_handler(FwIndexType portNum, Fw::Buffer& data, const ComCfg::FrameContext& context) {
    if !validateHeader() {
        recordRejection()
    }
    if !validateSequenceNumber() {
        recordRejection()
    }
    if !validateAuthKey() {
        recordRejection()
    }
    # process packet
}
```